### PR TITLE
Redis cache

### PR DESCRIPTION
**Pull request by NayHtetKyaw**

### CODE REVIEW

File: src/utils/redis.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    const results = await data?.toArray();
+
+    if (!results) {
+      console.log("There is no studyplan in Database.");
+      return;
+    }
+
+    const studyPlans = results.map((result) => convertResultToStudyPlan(result));
+
+    console.log('Cache Study Plans', studyPlans);
```
Comment (anascence) [inline src/utils/redis.ts]: not working at all

File: README.md [hunk 2]
```diff
@@ -30,6 +30,15 @@ Create a `.env` file with the following content
...
 S3_ENDPOINT=http://localhost:8333/
 MONGODB_URI=mongodb://localhost:27017
+REDIS_USERNAME=Chinathai
+REDIS_PASSWORD=1ULayo6WouIKTk1
+REDIS_HOST=103.214.23.122
+REDIS_PORT=6379
+REDIS_TLS=true
+REDIS_KEY_PATH=./redis_user_private.key
+REDIS_CERT_PATH=./redis_user_crt
+REDIS_CA_PATH=./redis_ca.pem
```
Comment (chinathaip) [inline README.md]: CRITICAL: please put example data for this and remove all the commits up to the one that this sensitive information has been introduced!

our repository is public, which means everybody can see this sensitive information

**Review by chinathaip — CHANGES_REQUESTED** [review]
I'll try have a look

for now please do the following XD

File: README.md (hunk 2 above)
Comment (anascence) [reply README.md]: ohh, i'll remove it

Comment (chinathaip) [reply README.md]: yep, simply remove these, and then squash all the commits into one I think

**Comment by chinathaip** [comment]
The secrets are still leaked in the commit history, can you squash all the commits into one? @NayHtetKyaw

**Comment by anascence** [comment]
> squash all the commits into one

i'll rebase, it is okay?

**Comment by chinathaip** [comment]
> > squash all the commits into one
> 
> i'll rebase, it is okay?

yep ok

**Comment by anascence** [comment]
> > > squash all the commits into one
> > 
> > 
> > i'll rebase, it is okay?
> 
> yep ok

is that okay? now ! @chinathaip

**Comment by chinathaip** [comment]
> > > > squash all the commits into one
> > > 
> > > 
> > > i'll rebase, it is okay?
> > 
> > 
> > yep ok
> 
> is that okay? now ! @chinathaip

no XDD there are 7 commits now, I can still go back to previous commits and see the secrets.

I'll rebase it for you no worries! And I'll also see what went wrong with the code and let you know

**Comment by anascence** [comment]
> > > > > squash all the commits into one
> > > > 
> > > > 
> > > > i'll rebase, it is okay?
> > > 
> > > 
> > > yep ok
> > 
> > 
> > is that okay? now ! @chinathaip
> 
> no XDD there are 7 commits now, I can still go back to previous commits and see the secrets.
> 
> I'll rebase it for you no worries! And I'll also see what went wrong with the code and let you know

5555 thank you krub! i am lost help.

File: src/app.ts [hunk 3]
```diff
@@ -3,16 +3,30 @@ import cors from "cors";
...
 import resourceRouter from "@routes/resources.route";
+import cacheEndpoint from "@middlewares/rediscache"; 
+import { getAllStudyPlans } from '@controllers/studyplans.controller';
 
 const app = express();
 app.get("/", (req: Request, res: Response) => {
 	res.send("Express + TypeScript Server");
 });
 
+app.get('/api/resources/study_plans', getAllStudyPlans);
```
Comment (chinathaip) [inline src/app.ts]: You can remove this line :) 

This thing is already done in 
`app.use("/api/resources", resourceRouter)`
`resourceRouter.get("/:page", getAllResources);`

The api simply fetches from the page name

https://github.com/stamford-syntax-club/stamfordcenter-backend/blob/main/src/routes/resources.route.ts#L6C1-L6C47

File: src/app.ts [hunk 4]
```diff
@@ -3,16 +3,30 @@ import cors from "cors";
...
 export default app;
+
+
+
+
+
+
+
+
+
```
Comment (chinathaip) [inline src/app.ts]: nit: can you remove white space for me? XD

File: src/controllers/studyplans.controller.ts [hunk 5]
```diff
@@ -1,22 +1,32 @@
...
+    const client = await getConnection();
+    const data = client?.db("stamfordcenter").collection("study_plans").find({});
+    const results = await data?.toArray();
+
+    if (!results) {
+      return res.status(404).send("file not found");
+    }
+
+    const studyPlans = results.map((result) => convertResultToStudyPlan(result));
+    res.locals.studyPlans = studyPlans;
```
Comment (chinathaip) [inline src/controllers/studyplans.controller.ts]: Ooh I never knew about setting a local variable within this scope. Thanks!
https://www.geeksforgeeks.org/express-js-res-locals-property/

File: src/middlewares/rediscache.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+            if (typeof req.originalUrl === 'string' && typeof value === 'string') {
+                await redisClient?.setEx(req.originalUrl, ttl, value);
+                console.log(`Data for endpoint: ${endpoint} cached for 30 days.`);
+            } else {
+                console.error('Invalid argument type for Redis setex command');
+            }
+        }else{
+            console.error(`No data found for endpoint: ${endpoint}`);
+        }
+    }
```
Comment (chinathaip) [inline src/middlewares/rediscache.ts]: didn't you already cache it in `studyplans.controller.ts`?

I think there is no need to do it again

File: src/app.ts [hunk 7]
```diff
@@ -3,16 +3,30 @@ import cors from "cors";
 import fileRouter from "@routes/files.route";
 import studyplanRouter from "@routes/studyplans.route";
 import resourceRouter from "@routes/resources.route";
+import cacheEndpoint from "@middlewares/rediscache"; 
+import { getAllStudyPlans } from '@controllers/studyplans.controller';
```
Comment (chinathaip) [inline src/app.ts]: following the removal of line 14 (`app.get('/api/resources/study_plans', getAllStudyPlans);`), this import can be removed too

**Review by chinathaip — CHANGES_REQUESTED** [review]
Looking gudddd, just a few points!

File: src/app.ts (hunk 3 above)
Comment (anascence) [reply src/app.ts]: i've tried without it, somehow it was not working, that's why i added that

File: src/middlewares/rediscache.ts (hunk 6 above)
Comment (anascence) [reply src/middlewares/rediscache.ts]: i think it's better to cache with this function, and can i just remove the other one in studyplan instead?

File: src/app.ts (hunk 3 above)
Comment (chinathaip) [reply src/app.ts]: lemme try

Comment (chinathaip) [reply src/app.ts]: I removed it and it "works" for me na  - the program attempts to get from redis before retrieving from database

File: src/middlewares/rediscache.ts (hunk 6 above)
Comment (chinathaip) [reply src/middlewares/rediscache.ts]: you will need to use `app.locals.XXX` to reuse the retrieved value in the middleware 

https://expressjs.com/en/4x/api.html#res.locals

File: src/middlewares/rediscache.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    const redisClient = await getRedisClient();
+    const cachedResponse = await redisClient?.get(req.originalUrl);
+
+    if (!cachedResponse) {
+        console.log(`No cache found for endpoint: ${endpoint}. Fetching data from database.`);
+        next(); // Get data from the database
+
+        if (res.locals.studyPlans) {
+            const value = JSON.stringify(res.locals.studyPlans);
+            if (typeof req.originalUrl === 'string' && typeof value === 'string') {
```
Comment (chinathaip) [inline src/middlewares/rediscache.ts]: `req.originalUrl` can be guaranteed to be of type string

`JSON.stringify()`  as its name suggests, it is guaranteed to convert your value to be of type string 

There is no need to check for these conditions na.

https://expressjs.com/en/4x/api.html#req.originalUrl
https://www.w3schools.com/js/js_json_stringify.asp

File: src/controllers/studyplans.controller.ts [hunk 9]
```diff
@@ -1,22 +1,32 @@
+import { Request, Response } from "express";
 import { convertResultToStudyPlan } from "@utils/converter";
 import { getConnection } from "@utils/mongoconnection";
-import { Request, Response } from "express";
+import { getRedisClient } from "@utils/redisclient";
+
+const cacheTTL = 60 * 24 * 30; // cache for 30 days
 
-// to be removed after frontend has migrated
```
Comment (chinathaip) [inline src/controllers/studyplans.controller.ts]: I just noticed you were making changes to `studyplans.controller.ts`. This file is deprecated, please use `resources.controllers.ts` instead

**Review by chinathaip — CHANGES_REQUESTED** [review]

File: src/controllers/studyplans.controller.ts (hunk 9 above)
Comment (anascence) [reply src/controllers/studyplans.controller.ts]: oh hai hai, i'll try to redo it with all the parts you mentioned

File: src/controllers/resources.controller.ts [hunk 10]
```diff
@@ -2,13 +2,18 @@ import { convertResultToQuickLink, convertResultToResource, convertResultToStudy
 import { getConnection } from "@utils/mongoconnection";
 import { Request, Response } from "express";
 import { Converters } from "@models/converters.model";
+import { getRedisClient } from "@utils/redisclient";
+import { get } from "http";
```
Comment (chinathaip) [inline src/controllers/resources.controller.ts]: Was it added by accident when you tried to type getAllResource?

File: src/app.ts [hunk 11]
```diff
@@ -3,16 +3,19 @@ import cors from "cors";
...
 const app = express();
 app.get("/", (req: Request, res: Response) => {
 	res.send("Express + TypeScript Server");
 });
 
 app.use(cors());
+app.use(cacheEndpoint); // Cache GET requests
 
 app.use("/api/study_plans", studyplanRouter); // to be removed after frontend has migrated
 app.use("/api/files", fileRouter);
```
Comment (chinathaip) [inline src/app.ts:18]: BTW have you tried retrieving files locally? I'm afraid it might get stuck in the cache thingy

File: src/app.ts (hunk 7 above)
Comment (chinathaip) [reply src/app.ts]: still not removed na @NayHtetKyaw

**Review by chinathaip — COMMENTED** [review]
Can you test retrieving the file locally? It might break because of redis

File: src/controllers/resources.controller.ts (hunk 10 above)
Comment (anascence) [reply src/controllers/resources.controller.ts]: ohh yea, it was an accident. i didn't know it was there.

File: src/app.ts (hunk 11 above)
Comment (anascence) [reply src/app.ts:18]: hai hai, let me try again.

**Review by chinathaip — APPROVED** [review]
LGTM, thank you for this PR  Mr.Anascence!

Feel free to merge
